ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoBackgroundController.swift

```diff
@@ -0,0 +1,364 @@
+      self?.cancelPlay(animated: true)
+    }
+
+    videoButtonsView.snp.makeConstraints {
+      $0.edges.equalToSuperview()
+    }
+  }
+
+  override func viewDidLayoutSubviews() {
+    super.viewDidLayoutSubviews()
+
+    self.playerLayer.frame = self.view.bounds
+
+    updatePlayerVisibility()
+  }
+
+  override func viewDidDisappear(_ animated: Bool) {
+    super.viewDidDisappear(animated)
+
+    // Cancel play if view becomes hidden.
+    cancelPlay(animated: false)
+  }
+
+  func startVideoPlayback() {
+    if !previewAutoplayFinished {
+      autoplayFinished()
+    }
+
+    self.videoButtonsView.isHidden = false
+
+    playerLayer.player?.isMuted = false
+    playerLayer.player?.seek(to: CMTime.zero)
+    playerLayer.player?.currentItem?.forwardPlaybackEndTime = CMTime()
+
+    let interval = CMTime(
+      seconds: 0.1,
+      preferredTimescale: CMTimeScale(NSEC_PER_SEC)
+    )
+    // Setup observer to check if 25% of the video is played.
+    timeObserver =
+      playerLayer.player?.addPeriodicTimeObserver(forInterval: interval, queue: nil) {
+        [weak self] time in
+        self?.checkPlayPercentage()
+      }
+
+    playerLayer.player?.play()
+    playStarted = true
+
+    playStartedEvent?()
+  }
+
+  private func setupPlayer(backgroundVideoPath: URL) {
+    self.view.backgroundColor = parseColorFromFilename(
+      filename: backgroundVideoPath.lastPathComponent
+    )
+    let stopFrame = parseStopFrameFromFilename(filename: backgroundVideoPath.lastPathComponent)
+    let resizeToFill = shouldResizeToFill(filename: backgroundVideoPath.lastPathComponent)
+
+    let asset: AVURLAsset = AVURLAsset(url: backgroundVideoPath)
+    loadVideoTrackParams(
+      asset: asset,
+      stopFrame: stopFrame,
+      resizeToFill: resizeToFill
+    )
+  }
+
+  private func loadVideoTrackParams(asset: AVURLAsset, stopFrame: Int?, resizeToFill: Bool) {
+    Task {
+      let isPlayable = try? await asset.load(.isPlayable)
+      let duration = try? await asset.load(.duration)
+      var frameRate: Float?
+      if let videoTrack = try? await asset.loadTracks(withMediaType: .video).first {
+        frameRate = try? await videoTrack.load(.nominalFrameRate)
+      }
+
+      guard let duration = duration,
+        let isPlayable = isPlayable
+      else {
+        videoLoaded(succeeded: false)
+        return
+      }
+      videoLoaded(succeeded: isPlayable)
+
+      setupPlayerLayer(
+        asset: asset,
+        resizeToFill: resizeToFill,
+        duration: duration,
+        frameRate: frameRate,
+        stopFrame: stopFrame
+      )
+    }
+  }
+
+  private func setupPlayerLayer(
+    asset: AVURLAsset,
+    resizeToFill: Bool,
+    duration: CMTime,
+    frameRate: Float?,
+    stopFrame: Int?
+  ) {
+    let item = AVPlayerItem(asset: asset)
+    let player = AVPlayer(playerItem: item)
+    playerLayer.player = player
+
+    if resizeToFill {
+      playerLayer.videoGravity = .resizeAspectFill
```

Comment:
This function is being called on a `Task` that is not main-actor isolated. You should not adjust any UI on a non-main isolated thread.

Reply:
Actually `Task` is called on main actor because:
`NewTabPageVideoBackgroundController` is a child of UIViewController class so it is attached to the MainActor, so the default actor of our `Task` is the MainActor.

I can specify `in MainActor` explicitly for a Task for readability.